### PR TITLE
Add Deep Potential plugin for CP2K

### PR DESCRIPTION
**Pull request by Cloudac7**
Deep Potential (DP) is a machine learning potential based on deep neural network, first proposed by Zhang L. et al. (Doi: [10.1016/j.cpc.2018.03.016](https://doi.org/10.1016/j.cpc.2018.03.016)) By learning from coordinates as well as energy and forces of different structures from DFT calculations, a potential describing the whole system could be frozen to a file. By reading the file, program could yield energy, forces etc. from structures, and then MD simulation with DFT accuracy could be fastened greatly. This PR add Deep Potential support for CP2K, with potential file (e.g. `graph.pb`) supplied, CP2K could run MD simulation. 

**Only DeePMD-kit >= 2.0.0 is supportted.**

To enable the DeePMD-kit feature, `--with-deepmd` as well as `--with-tfcc` flag in toolchain should not be set to `no`. The following ways installing has been added.

 - Please set `--deepmd-mode` to `gpu` if GPU support is needed, and load CUDA environment before installing.
 - From exist DeePMD-kit C++ interface, with`$deepmd_root` and `$tensorflow_root` (Described [here](https://docs.deepmodeling.org/projects/deepmd/en/latest/install.html#install-the-deepmd-kit-s-c-interface) in detail), by setting `--with-deepmd=$deepmd_root` and `--with-tfcc=$tensorflow_root`.
  - From source code and library from Github and Anaconda Cloud, by setting the both flags to `install`. For now, DeePMD-kit v2.0.0.b1 and Tensorflow C++ interface 2.3.0 will be installed, and only CUDA Toolkit 10.1 is OK from this way. 
  - From system, by adding the header files and library to where `ld` command could read directly.

The result from CP2K has been tested to be the same with which from Lammps, supportted by Deepmodeling.

### CODE REVIEW

**Comment by oschuett** [comment]
Thanks for the PR. This is certainly impressive work!

As a first step, I'd ask you to remove any merge commits from your git history, ie. make it linear.

Furthermore, I'd suggest you squash the git history to a single (or few) commit(s).

File: tools/toolchain/scripts/stage6/install_deepmd.sh [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    ;;
+  __INSTALL__)
+    echo "==================== Installing TensorFlow C++ Interface ===================="
+    pkg_install_dir="${INSTALLDIR}/deepmd-kit-${deepmd_ver}"
+    export DP_VARIANT=${DEEPMD_MODE}
+    if [ -f libtensorflow_cc-${tfcc_ver}.tar.bz2 ]; then
+      echo "libtensorflow_cc-${tfcc_ver}.tar.bz2 is found"
+    else
+      download_pkg ${DOWNLOADER_FLAGS} ${deepmd_sha256} \
+        https://anaconda.org/deepmodeling/libtensorflow_cc/2.3.0/download/linux-64/libtensorflow_cc-${tfcc_ver}.tar.bz2
```
Comment (oschuett) [inline tools/toolchain/scripts/stage6/install_deepmd.sh:58]: Installing Tensorflow via Conda is a nice solution given how painful it is to build from source. However, I'm wondering why are you not using the "official" package from [conda forge](https://anaconda.org/conda-forge/libtensorflow_cc)?

File: tools/toolchain/scripts/stage6/install_deepmd.sh [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    tfcc_ver="2.3.0-cpu_cudaNone_0"
+    tfcc_sha256="b1b6e5c68a58e516178e6e855a70dd020021d1cf22a94ad291bb8e7541ec58db"
+    ;;
+  cuda)
+    tfcc_ver="2.3.0-gpu_cuda10.1_0"
+    tfcc_sha256="085ba903c76adbbcb57dcdb474e5da5fd36324ca64ce673acacf122d5a2f2b3d"
+    ;;
+esac
+
+cub_link="https://github.com/NVlabs/cub/archive/c3cceac115c072fb63df1836ff46d8c60d9eb304.zip"
```
Comment (oschuett) [inline tools/toolchain/scripts/stage6/install_deepmd.sh:26]: I believe CUB is now [included](https://docs.nvidia.com/cuda/cub/index.htm) in the regular CUDA Toolkit. So, there is no need to install it separately anymore.

File: src/force_fields_input.F [hunk 3]
```diff
@@ -612,6 +622,42 @@ SUBROUTINE read_eam_section(nonbonded, section, start, para_env, mm_section)
       END DO
    END SUBROUTINE read_eam_section
 
+! **************************************************************************************************
+!> \brief Reads the DEEPMD section
+!> \param nonbonded ...
+!> \param section ...
+!> \param start ...
+!> \author teo
```
Comment (oschuett) [inline src/force_fields_input.F:630]: You can put your own name here or remove the `\author` tag as you prefer.

File: src/manybody_deepmd.F [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+! change to pointer type
+      dener => pot_deepmd
+      dforce => dpmd_force
+      dvirial => dpmd_virial
+      datom_ener => dpmd_atom_ener
+      datom_virial => dpmd_atom_virial
+      dcoord => dpmd_coord
+      dbox => dpmd_box
+      datype => dpmd_atype
+!      IF (do_parallel) THEN
```
Comment (oschuett) [inline src/manybody_deepmd.F:176]: Left overs copied from QUIP? See also remainder of this routine.

File: tools/toolchain/scripts/arch_base.tmpl [hunk 5]
```diff
@@ -10,6 +10,7 @@ WFLAGS      = ${WFLAGS}
 #
 FCDEBFLAGS  = ${FCDEBFLAGS}
 CFLAGS      = ${CFLAGS}
+CXXFLAGS    = ${CXXFLAGS}
```
Comment (oschuett) [inline tools/toolchain/scripts/arch_base.tmpl:13]: Does DeePMD-kit also offer a C API or would it be easy to add? So, far we've managed to keep C++ out of CP2K and I'd really like to keep it that way.

**Review by oschuett — COMMENTED** [review]
Thanks for fixing the git history. I left a couple of comments. Once everything is up and running it would also be good to add some regression tests.

File: src/deepmd_c_wrapper.h [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+/*  Copyright 2000-2021 CP2K developers group <https://cp2k.org>              */
+/*                                                                            */
+/*  SPDX-License-Identifier: GPL-2.0-or-later                                 */
+/*----------------------------------------------------------------------------*/
+
+/*******************************************************************************
+ * \brief A minimal wrapper for DeepMD-kit C++ interface.
+ * \author Yongbin Zhuang and Yunpei Liu
+ ******************************************************************************/
+
```
Comment (dev-zero) [inline src/deepmd_c_wrapper.h:12]: include guard `#ifndef DEEPMOD_C_WRAPPER_H ...` missing

File: src/deepmd_wrapper.F [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+      REAL(C_DOUBLE), POINTER :: dvirial(:)
+      REAL(C_DOUBLE), POINTER :: datom_ener(:)
+      REAL(C_DOUBLE), POINTER :: datom_virial(:)
+      REAL(C_DOUBLE), POINTER :: dcoord(:)
+      INTEGER(C_INT), POINTER :: datype(:)
+      REAL(C_DOUBLE), POINTER :: dbox(:)
+#if(__DEEPMD)
+      CALL compute_nnp_c(pot, C_LOC(vecsize), C_LOC(dener), C_LOC(dforce(1)), &
+                         C_LOC(dvirial(1)), C_LOC(datom_ener(1)), C_LOC(datom_virial(1)), &
+                         C_LOC(dcoord(1)), C_LOC(datype(1)), C_LOC(dbox(1)))
```
Comment (dev-zero) [inline src/deepmd_wrapper.F:111]: Wouldn't `C_LOC` work on the arrays directly rather than taking the first argument first?
Also, it should be possible to drop `POINTER` but they should likely to be declared `CONTIGUOUS` instead.

File: src/manybody_deepmd.F [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+                                                  output_unit
+      LOGICAL                                  :: do_parallel
+      LOGICAL, ALLOCATABLE                     :: use_atom(:)
+      REAL(kind=dp)                            :: lattice(3, 3)
+      TYPE(pair_potential_single_type), &
+         POINTER                                :: pot
+      TYPE(deepmd_data_type), POINTER            :: deepmd_data
+      TYPE(deepmd_pot_type), POINTER             :: deepmd
+      TYPE(nnp), SAVE                                :: deepmd_tmp_pot1, &
+                                                        deepmd_tmp_pot2
```
Comment (dev-zero) [inline src/manybody_deepmd.F:85]: No. If you have to preserve data you have to add it to a suitable env. No global vars or shared state.

File: src/manybody_deepmd.F [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+                             compute_nnp
+#endif
+
+#include "./base/base_uses.f90"
+
+   IMPLICIT NONE
+
+   PRIVATE
+   LOGICAL, SAVE, PRIVATE :: do_create(2) = .TRUE.
+   INTEGER, SAVE, PRIVATE :: do_count = 0
```
Comment (dev-zero) [inline src/manybody_deepmd.F:41]: No, see further down below.

**Review by dev-zero — CHANGES_REQUESTED** [review]

**Comment by Cloudac7** [comment]
Thanks for reviewing. We will check our code and modify then.

File: tools/toolchain/scripts/stage6/install_deepmd.sh (hunk 2 above)
Comment (Cloudac7) [reply tools/toolchain/scripts/stage6/install_deepmd.sh:26]: Well, in [DeePMD-kit repository](https://github.com/deepmodeling/deepmd-kit), they just use [the very fork](https://github.com/NVIDIA/cub/tree/b229817e3963fc942c7cc2c61715a6b2b2c49bed) of CUB as submodule, while removing it error would be raised when compiling. So here I tried to add it to DeePMD-kit code by downloading code archive file directly. According to your description, it might be OK building without it, and I will have a try.

Comment (Cloudac7) [reply tools/toolchain/scripts/stage6/install_deepmd.sh:26]: > I believe CUB is now [included](https://docs.nvidia.com/cuda/cub/index.htm) in the regular CUDA Toolkit. So, there is no need to install it separately anymore.

In their latest commits, they remove extra CUB requirement for CUDA>=11, shown [here](https://github.com/deepmodeling/deepmd-kit/commit/9b5e2b9b4d1c1ad53b1e6887177aab186fe412cf#diff-735c1a60650597d4822c37e3c7ea5157250a5baea1403120e26f73973ff6630c), as it has been included in CUDA Toolkit. However, for former CUDA versions, it might still be needed.

File: tools/toolchain/scripts/stage6/install_deepmd.sh (hunk 1 above)
Comment (Cloudac7) [reply tools/toolchain/scripts/stage6/install_deepmd.sh:58]: > Installing Tensorflow via Conda is a nice solution given how painful it is to build from source. However, I'm wondering why are you not using the "official" package from [conda forge](https://anaconda.org/conda-forge/libtensorflow_cc)?

The packages used here are provided by Deep Modeling, the main developer group of DeePMD-kit. I will also try the package on `conda-forge`.

Comment (Cloudac7) [reply tools/toolchain/scripts/stage6/install_deepmd.sh:58]: > The packages used here are provided by Deep Modeling, the main developer group of DeePMD-kit. I will also try the package on `conda-forge`.

Failed because of the lost of `include/google/protobuf/type.pb.h`. The package on `conda-forge` seems not include some headers needed by DeePMD-kit.

Comment (oschuett) [reply tools/toolchain/scripts/stage6/install_deepmd.sh:58]: That header file is probably part of [libprotobuf](https://github.com/conda-forge/tensorflow-feedstock/blob/f18fb006b85678a5180336e5f99f68a9ba20d5ea/recipe/meta.yaml#L60). So, I guess installing Conda packages by hand is not a very robust solution after all.

I think, then we should keep Tensorflow out of the toolchain for now. It's actually not that hard to install via pip and most HPC centers also provide it.

I've opened #1622 to make Tensorflow available in the CI. Once it's merged you can rebase this PR and continue testing.

File: tools/toolchain/scripts/stage6/install_deepmd.sh (hunk 2 above)
Comment (oschuett) [reply tools/toolchain/scripts/stage6/install_deepmd.sh:26]: Ok, then let's keep it for a year or so until everybody has upgrade to CUDA >= 11.

**Comment by oschuett** [comment]
Tensorflow is now readily available within the CI containers (#1622). So, please rebase your branch and adopt `install_deepmd.sh`.

**Comment by oschuett** [comment]
Gentle ping, is this PR still being worked on?

**Comment by oschuett** [comment]
@deepmodeling, does anybody know if there is still any interest in this PR?

**Comment by oschuett** [comment]
I'm going to close this PR now. Personally, I think it's a really pity because obviously a lot of work went into it. But I'm afraid this wouldn't be useful to anybody without further support from the authors.

**Comment by robinzyb** [comment]
> I'm going to close this PR now. Personally, I think it's a really pity because obviously a lot of work went into it. But I'm afraid this wouldn't be useful to anybody without further support from the authors.

hi, @oschuett,
    @Cloudac7 and me are the contributors to this pr. I will make a new pr since my colleagues have developed relevant FEP methods based on DPMD(https://aip.scitation.org/doi/10.1063/5.0098330) and we  are planning to reorganize these codes in a more pretty format.

**Comment by robinzyb** [comment]
@oschuett is it ok for opening a new subsection named  DPNN/DPMD under FORCE_EVAL or nest the section in FORCE_EVAL/NNP

**Comment by oschuett** [comment]
Hi @robinzyb, this is great news!

>  is it ok for opening a new subsection named DPNN/DPMD under FORCE_EVAL or nest the section in 

Yes, please create new subsections below [FORCE_EVAL](https://manual.cp2k.org/trunk/CP2K_INPUT/FORCE_EVAL.html), [MOTION/MD](https://manual.cp2k.org/trunk/CP2K_INPUT/MOTION/MD.html), and [MOTION/FREE_ENERGY](https://manual.cp2k.org/trunk/CP2K_INPUT/MOTION/FREE_ENERGY.html) as you see fit.

**Comment by VaradRengaraj** [comment]
> @oschuett is it ok for opening a new subsection named DPNN/DPMD under FORCE_EVAL or nest the section in FORCE_EVAL/NNP


> Hi @robinzyb, this is great news!
> 
> > is it ok for opening a new subsection named DPNN/DPMD under FORCE_EVAL or nest the section in
> 
> Yes, please create new subsections below [FORCE_EVAL](https://manual.cp2k.org/trunk/CP2K_INPUT/FORCE_EVAL.html), [MOTION/MD](https://manual.cp2k.org/trunk/CP2K_INPUT/MOTION/MD.html), and [MOTION/FREE_ENERGY](https://manual.cp2k.org/trunk/CP2K_INPUT/MOTION/FREE_ENERGY.html) as you see fit.


Hi @robinzyb , i was told to work on this PR. Can you please confirm if you are already working on this?
